### PR TITLE
Add UI for bypassing SSL handshake failures

### LINKED ISSUE #26683: Make it possible to accept self-signed certs
There are two main issues at the moment:
* the embedding API embeds the contents of resources/certs into the binary (https://github.com/servo/servo/blob/748b4241b7f2b7be7dc71be4df1ad89ddec49d99/ports/libsimpleservo/api/src/lib.rs#L926)
* we need to recognize when an unsupported cert causes the connection to be terminated and offer a UI that can bypass that error if desired

### PR DESCRIPTION
**Pull request by jdm**
There are several parts to these changes:
1. resurrecting the network error classification code to distinguish between SSL failures and other network errors
1. adding an SSL verification callback to support verifying certs against a list that can change at runtime, rather than just at program initialization
1. exposing a privileged chrome://allowcert URI which accepts the PEM cert contents along with a secret token
1. extracting the PEM cert contents out of the network layer when a handshake failure occurs, and getting them into the HTML that is parsed when an SSL failure occurs
1. adding a button in the handshake failure page that performs an XHR to chrome://allowcert with knowledge of the secret token and the PEM cert contents, before reloading the original URL that failed

The presence of the secret token means that while the chrome://allowcert URL is currently visible to web content, they cannot make use of it to inject arbitrary certs into the verification process.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #26683
- [x] These changes do not require tests because the UI requires user activation and can't clearly be automated

### CODE REVIEW

**Comment by jdm** [comment]
Screen Shot 2020-05-29 at 1 35 56 PM

Here's an example of loading the default Hubs local dev server with the new UI. Pressing the button reloads the page without a handshake error.

File: components/net/connector.rs [hunk 1]
```diff
@@ -30,6 +35,38 @@ const SIGNATURE_ALGORITHMS: &'static str = concat!(
...
+    pub(crate) fn new() -> Self {
+        Self {
+            certs: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    fn store(&self, host: String, cert_bytes: Vec<u8>) {
+        let mut certs = self.certs.lock().unwrap();
+        let entry = certs.entry(host).or_insert((cert_bytes, 0));
+        entry.1 += 1;
```
Comment (asajeffrey) [inline components/net/connector.rs:53]: This looks a lot like a poor dev's weak Rc. Is it?

File: resources/badcert.html [hunk 2]
```diff
@@ -3,6 +3,24 @@
 <title>Certificate error</title>
 </head>
 <body>
-    <p>${reason}</p>
+  <p>${reason}</p>
+  <pre id="bytes">${bytes}</pre>
+  <button id="allow">Allow certificate temporarily</button>
```
Comment (asajeffrey) [inline resources/badcert.html:8]: Should there also be a "decline" button?

File: resources/badcert.html [hunk 3]
```diff
@@ -3,6 +3,24 @@
...
+  <script>
+    let bytes = document.getElementById('bytes').textContent;
+    let button = document.getElementById('allow');
+    if (bytes.length) {
+        button.onclick = function() {
+            let xhr = new XMLHttpRequest();
+            xhr.open('GET', 'chrome:allowcert?secret=${secret}&bytes=' + btoa(bytes));
```
Comment (asajeffrey) [inline resources/badcert.html:15]: I'm a bit concerned that we're using a GET request for this, side-effecting GET that adds a security wormhole scares me!

**Review by asajeffrey — COMMENTED** [review]
This all looks like a sensible approach, apart from using a side-effecting GET as the way of bypassing cert checks.

File: resources/badcert.html (hunk 2 above)
Comment (jdm) [reply resources/badcert.html:8]: Perhaps. What should it do, however?

File: resources/badcert.html (hunk 3 above)
Comment (jdm) [reply resources/badcert.html:15]: What is the security wormhole, exactly?

**Comment by jdm** [comment]
Note to self: unit tests don't build now.

File: components/net/connector.rs (hunk 1 above)
Comment (jdm) [reply components/net/connector.rs:53]: It does have the property of removing the hashmap entry when the last user goes away, so it's slightly more complex than Rc/Arc.

File: resources/badcert.html (hunk 2 above)
Comment (jdm) [reply resources/badcert.html:8]: Firefox does have the Get Me Out Of Here button which I think does the same thing, so I'll add that.

File: components/net/connector.rs (hunk 1 above)
Comment (Manishearth) [reply components/net/connector.rs:53]: Yeah I could go in either direction here

File: resources/badcert.html (hunk 3 above)
Comment (Manishearth) [reply resources/badcert.html:15]: Yeah IMO this should be a POST. Secrets in url params is bad

**Review by Manishearth — APPROVED** [review]
r=me with minor changes, and we should also be passng unit tests

**Comment by jdm** [comment]
Fixed the unit tests, added on for accepting self-signed certs, and added a button to go back in history on the cert error page.
@bors-servo r=Manishearth

**Comment by jdm** [comment]
@bors-servo retry
#25513

**Pull request merged**